### ISSUE TITLE
clang-tidy feature request: C++ Core Guidelines "ES.23: Prefer the `{}`-initializer syntax" (avoid `()` initialization)

### ISSUE THREAD

**Issue by N-Dekker** [llvm/llvm-project#61454]
C++ Core Guidelines, September 23, 2022, item [ES.23: Prefer the `{}`-initializer syntax](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-list) says:

> Prefer {}. The rules for {} initialization are simpler, more general, less ambiguous, and safer than for other forms of initialization.
> 
> Use = only when you are sure that there can be no narrowing conversions. For built-in arithmetic types, use = only with auto.
>
> Avoid () initialization, which allows parsing ambiguities.

So for example, when declaring an `m` x `n` matrix object (assuming a `Matrix::Matrix(int m, int n)` constructor):

```cpp
Matrix oldStyleMatrix(m, n);  // Old style: initialization using parentheses
Matrix newStyleMatrix{m, n};  // New style: initialization using curly braces.
```

The new style (using curly braces) appears preferable, according to this guideline. Would it be possible to add a clang-tidy check, to warn against using the old style, as well as the ability to convert code to the new style (using the option `-fix`)?

**Comment by samacumen** [comment]
Are C++ core guidelines fully supported in all of clang tools? I see it is with clang-tidy and clang-format. Any other reference would help.

**Comment by JonasToth** [comment]
> Are C++ core guidelines fully supported in all of clang tools? I see it is with clang-tidy and clang-format. Any other reference would help.

No. The cppcoreguidelines are not fully supported. There is no "global preset" or anything as an best-effort available either.
`clang-tidy` has its module and probably a few checks of other modules address issues of the coreguidelines as well (they are often aliased into the `cppcoreguidelines` module too).

**Comment by vbvictor** [comment]
I'll try to give a go for this check.

Currently I', aiming at _not_ providing warn/fixits for following cases:
1. Overloaded ctor that accept `std::initializer_list`
2. Inside template context (can change behavoir because of N.1)
3. `auto` variables (same as N.1)

Appreciate if you know any more corner cases that should be carefully considered.
